src/Internal/Payment/PaymentRequest.php

```diff
@@ -144,4 +144,19 @@ public function get_payment_method(): PaymentMethodInterface {
 
 		throw new PaymentRequestException( __( 'No valid payment method was selected.', 'woocommerce-payments' ) );
 	}
+
+	/**
+	 * Populates a payment context before processing a payment.
+	 *
+	 * This method is the link between the payment request, and the payment process.
+	 * Use it to make sure that all necessary parameters are provided in advance,
+	 * or throw an exception otherwise. Once done, the payment process would rely
+	 * on all needed parameters being in place.
+	 *
+	 * @param PaymentContext $context Context to populate.
+	 * @throws PaymentRequestException When data is not available or invalid.
+	 */
+	public function populate_context( PaymentContext $context ) {
```

Comment:
This method here does not feel right to me. I don't think `PaymentRequest` needs to be aware or act on anything related to `PaymentContext` in its own class.

IMO, we would move this method to `PaymentProcessingService`, where it interacts with all relevant payment stuff: states (in a state machine), PaymentContext, and PaymentRequest. We may try to separate these stuff as much as possible.

Reply:
@htdat please check this conversation: https://github.com/Automattic/woocommerce-payments/pull/7332#discussion_r1342767258

I'll be moving this context population to `InitialState` to keep all context population in the same place.

Reply:
Ah, I did not notice that conversation. I will follow up in that convo.